长期预约页面跳转的不良体验修复

jiaxinou2022 opened the pull request:
主要修改了Appointment/view.py/checkout_appoint函数，把一部分参数不合法的校验由POST之前放到POST过程中，并用wrong()作为弹窗抛出，避免了页面直接跳转回主页面的不良体验

jiaxinou2022 committed:
干净多了

Deophius commented:
Closes #854

Deophius commented on Appointment/views.py:
```diff
@@ -424,6 +424,11 @@ def arrange_time(request: HttpRequest):
     is_person = request.user.is_person()
 
+     # 处理从URL参数传递过来的错误信息
+    html_display = {}
```
这个地方加了 `html_display`，但是 template 里面并没有处理这个的代码，所以好像没有把错误信息显示出来。可以在长期预约页面的“预约周数”输入"asdf"观察到这个bug。

Deophius commented on Appointment/views.py:
```diff
@@ -807,9 +810,17 @@ def checkout_appoint(request: UserRequest):
     except AssertionError as e:
-        return redirect(message_url(wrong(f'参数不合法: {e}'), reverse('Appointment:index')))
+        # 参数不合法时，重定向回arrange_time页面并显示错误信息
+        redirect_url = f'/underground/arrange_time?Rid={Rid}&start_week={start_week}'
```
这里 `Rid` 好像是直接从用户的输入里面获得的，而且没有检查过？个人感觉最好检查一下用户输入的合法性

Deophius replied:
`start_week` 也是

jiaxinou2022 replied:
这一点我有些疑问，Rid和start_week不是手动输入的，都是用户通过按钮选出来的，暂时想不到有什么情况下能有非法输入啊？

Deophius commented on Appointment/views.py:
```diff
@@ -807,9 +810,17 @@ def checkout_appoint(request: UserRequest):
+        if is_longterm:
+            redirect_url += '&longterm=on'
+        return redirect(message_url(wrong(f'参数不合法: {e}'), redirect_url))
     except:
```
这里似乎可以加一个`except ValueError`，用来处理上面几个 `int()` 调用失败的情况？

Deophius commented on Appointment/views.py:
```diff
@@ -883,9 +894,14 @@ def checkout_appoint(request: UserRequest):
+        if is_longterm and not times:
+            wrong("长期预约周数未填写", render_context)
+        # 检查间隔周数
+        elif is_longterm and not (1 <= interval <= CONFIG.longterm_max_interval):
```
感觉是不是用 `if` 就行了

Deophius commented on Appointment/views.py:
```diff
@@ -1073,4 +1089,4 @@ def review(request: HttpRequest):
 def logout(request):    # 登出系统
     auth.logout(request)
-    return redirect(reverse('Appointment:index'))
+    return redirect(reverse('Appointment:index'))
```
文件末尾的空行还是留着吧，好像大部分文件都留着

jiaxinou2022 committed:
微调

补全了booking.html的内容，加入了ValueError的部分，针对预约周数非数字的情况进行了修复，修复了elif的部分。
此外关于Rid和start_week的合法性检验暂时没有加入。

MinionXX611 commented:
已加入`Rid`和`start_week`的合法性检验。
达到的效果是：直接向网页发送不合法信息时，会跳转到预约主页并展示错误信息：`非法的预约信息！`

Deophius commented on Appointment/views.py:
```diff
@@ -860,6 +890,16 @@ def checkout_appoint(request: UserRequest):
+                wrong("长期预约周数或间隔周数格式错误，请输入数字", render_context)
+                # 简化错误处理
+                render_context.update(contents=contents, show_clause=True)
+                return render(request, 'Appointment/checkout.html', render_context)
```
这里 `render` 似乎会导致后面的 `search_users` 等变量没有进入 `render_context`。这样在显示错误信息的页面无法选择参与预约的成员

jiaxinou2022 committed:
进一步修复

修改了rid和start_week的合法性检验，同时解决了在显示错误信息的页面无法选择参与预约的成员的问题

Morgen-Kornblume reviewed (approved):
Checked in dev env and approve to merge.

Morgen-Kornblume committed:
Merge branch 'develop' into yuyue

Morgen-Kornblume commented on Appointment/views.py:
```diff
@@ -787,29 +795,61 @@ def checkout_appoint(request: UserRequest):
     except AssertionError as e:
-        return redirect(message_url(wrong(f'参数不合法: {e}'), reverse('Appointment:index')))
+        # 参数不合法时，重定向回arrange_time页面并显示错误信息
+        safe_rid = room.Rid if room else "B107A"
```
此处重定向不应使用硬编码跳转回某个房间，正常入口是不会触发选择不存在的房间然后跳转B107A的，既然用户试图通过构造来找我们的漏洞，那么我们就不要管他的体验，直接跳转回主页让他睁开眼睛好好看看哪些房间是可以预约的！

Morgen-Kornblume replied:
上述针对的是用户给的RID不存在的情况跳转回主页，正常情况就继续使用正常的RID即可

Morgen-Kornblume reviewed (changes requested):
针对上述问题需要做出改进

Morgen-Kornblume commented on Appointment/views.py:
```diff
@@ -787,29 +795,61 @@ def checkout_appoint(request: UserRequest):
-        return redirect(message_url(wrong(f'参数不合法: {e}'), reverse('Appointment:index')))
+        # 参数不合法时，重定向回arrange_time页面并显示错误信息
+        safe_rid = room.Rid if room else "B107A"
+        redirect_url = f'/underground/arrange_time?Rid={safe_rid}&start_week={safe_start_week}'
```
另一个bug是，即使在上面判断了房间号不可预约，但是如果我输入不可预约房间的RID，此处跳转依然会直接跳转到该不可预约房间的页面（如图所示）。
<img width="2874" height="1704" alt="image" src="https://github.com/user-attachments/assets/d11296f2-6a91-4889-acb7-c0cff19b733b" />
如果RID是不可预约的房间，那么也应该跳转回主页……

MinionXX611 committed:
修复了房间预约参数合法性处理的不完善之处

MinionXX611 commented:
修改如下：

将`Rid`和`start_week`的合法性检验挪到最前面，以保证它们最先被处理，一旦错误（非法访问）就重定向到主页；其他参数不合法时（合法访问，但可能用户漏填信息），则跳转到预约界面，以保证用户体验。

将创建预约时的错误处理改为`重定向到主页`，主要是因为存在`房间已被预约但预约尚在审核中`的状态，此时房间不应该被预约，但是room模型中似乎没有对应的状态，我没能找到比较好的判断方式。然而在创建预约的时候已有处理这种情况的代码，所以我想直接利用这段代码拦截。由于前面已经对用户输入的参数进行过合法性检验，于是这里出现的错误只能是由非法访问造成，所以统一都重定向到主页处理。

Morgen-Kornblume committed:
调整跳转的“主页”

之前 review comment 没有表述清楚，“主页”指的是地下室预约界面的主页

Deophius committed:
Merge branch 'develop' into yuyue
